locator/tablets.cc

```diff
@@ -245,9 +246,42 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
     }
 }
 
-host_id tablet_map::get_primary_replica(tablet_id id) const {
+static const tablet_replica& sort_and_get_primary_replica(tablet_replica_set& replicas, tablet_id id) {
+    size_t size = replicas.size();
+    size_t idx = id.value() % size;
+    // partially sort as few items as possible
+    // although partial_sort doesn't guarantee stable sort of equal items
+    // we compare here host_id:s that must be unique, so there is no problem
+    // using std:;partial_sort
+    if (idx < size / 2) {
+        // sort the first idx replicas in ascending order
+        std::partial_sort(replicas.begin(), replicas.begin() + idx + 1, replicas.end(), [] (const tablet_replica& l, const tablet_replica& r) {
+            return l.host < r.host;
+        });
+    } else {
+        // sort the last (size - idx) replicas in descending order
+        std::partial_sort(replicas.rbegin(), replicas.rbegin() + size - idx + 1, replicas.rend(), [] (const tablet_replica& l, const tablet_replica& r) {
```

Comment:
I think this is incorrect. 

We need to sort either
[0, idx] (first part) or [idx, size-1] (last part).

For [idx, size -1 ], we make sure the last n = size -1 - idx + 1 = size - idx elements are sorted.

The code above sorts the last size - idx + 1 + 1 = size -  idx + 2 elements, which is wrong. 

It should be std::partial_sort(replicas.rbegin(), replicas.rbegin() + size - idx  -1, replicas.rend() ...

Reply:
hmm, I think your calculation is off by one, to the other (unsafe) side.

partial_sort sorts `[begin, middle)`, so when sorting the vector tail, say size=6, idx=3:

|  | 0 | 1 | 2 | 3 | 4 | 5 |
|---|---|---|---|---|---|---|
| rend |   |   |  mid |   |  | rbegin |

we want mid = rbegin + 3 = rbegin + (size - idx)

If idx=5:

|  | 0 | 1 | 2 | 3 | 4 | 5 |
|---|---|---|---|---|---|---|
| rend |   |   |    |   | mid | rbegin |

we want mid = rbegin + 1 = rbegin + (size - idx)

Reply:
Correct, since (rbegin + midlle  - rbegin  + 1  - 1 = size - idx) , so  middle  = size -idx for the reverse sort.